syntax = "proto3";

package tsdb.proto;

option go_package = "github.com/vobbilis/codegen/mytsdb/go-tsdb/proto/gen";

message Sample {
  int64 timestamp = 1;
  double value = 2;
}

message Label {
  string name = 1;
  string value = 2;
}

message TimeSeries {
  repeated Label labels = 1;
  repeated Sample samples = 2;
}

message LabelMatcher {
  enum Type {
    EQ = 0;
    NEQ = 1;
    RE = 2;
    NRE = 3;
  }
  Type type = 1;
  string name = 2;
  string value = 3;
}

message TimeRange {
  int64 start_time = 1;
  int64 end_time = 2;
}

message QueryParams {
  TimeRange time_range = 1;
  repeated LabelMatcher matchers = 2;
}

message LabelValuesRequest {
  string label_name = 1;
  QueryParams params = 2;
}

message LabelNamesResponse {
  repeated string names = 1;
}

message LabelValuesResponse {
  repeated string values = 1;
}

message SeriesResponse {
  repeated TimeSeries series = 1;
}

service TSDBService {
  rpc GetLabelNames(QueryParams) returns (LabelNamesResponse);
  rpc GetLabelValues(LabelValuesRequest) returns (LabelValuesResponse);
  rpc GetSeries(QueryParams) returns (SeriesResponse);
}
